models/optional_modules/consent/snowplow_web_consent_users.sql

```diff
@@ -0,0 +1,148 @@
+{{
+  config(
+    materialized='incremental',
+    unique_key='domain_userid',
+    sort = 'last_consent_event_tstamp',
+    dist = 'domain_userid',
+    sql_header=snowplow_utils.set_query_tag(var('snowplow__query_tag', 'snowplow_dbt'))
+  )
+}}
+
+with base as (
+
+  select
+    domain_userid,
+    user_id,
+    geo_country,
+    derived_tstamp as last_processed_event,
+    row_number() over (partition by domain_userid order by derived_tstamp desc) as row_count
+
+  from {{ ref('snowplow_web_consent_log') }}
+
+  where
+  -- enough time has passed since the event happened (derived_tstamp) to be able to process it as a whole (please bear in mind the late arriving data)
+  derived_tstamp < cast({{ dbt_utils.dateadd('hour', var("snowplow__consent_late_arriving_window", 10), dbt_utils.current_timestamp_in_utc() ) }} as {{ dbt_utils.type_timestamp() }})
+
+  {% if is_incremental() %} -- and it has not been processed yet
```

Comment:
❓ I'm not sure I quite understand this - are we saying we don't want to process consent events that were sent within the last 10 hours? Why would we want to do that? Also from a naming point of view, compared to `snowplow__days_late_allowed` which increases the window we look at this decreases it (I think?), so maybe a name like `snowplow__consent_delay_analysis_window_hours` or something might be clearer about the difference?

Also is there a reason you aren't using this in the `models/optional_modules/consent/snowplow_web_consent_cmp_stats.sql` model?

Reply:
So the idea behind this was that we wait until we are fine to exclude late arriving data and only process things once, otherwise we won't be able to sum up the aggregates. The same concept is in the media player package and my heart just stopped as I believe the logic is wrong, that's why you were confused, too. It should be -10, I believe, and it is currently wrongly calculated within the MP package (aka the Pageview is processed even if it is not yet ready) but please tell me I'm wrong here..

Reply:
I don't think you are wrong...

Reply:
Worth having someone else triple check!

Reply:
I could actually see it taking effect with the latest snowflake data coming in. I got the data around mid-day and was surprised to see data being modelled as 10 hours have not passed since we got the data. After I changed it to -10 it was not processed and with -1 I was able to process them.. :(

Reply:
On that note, I think we should probably change the logic instead of making this value a negative (aka adding the hours to the derived_tstamp) so that the variable naming make sense to users.

Reply:
Yep agree with that, I still think something like `delay` or `buffer` might make it more obvious and distinct to the variable we already have with late, but I'm not too fussed and would rather just align with the media player

Reply:
Renamed it to buffer, the media player has a specific convention due to it processing full pageview events so we are free to rename it as we wish tbf. Just noticed you had another question on why I did not make the cmp_stats incremental. It's cause of the weird lag/lead window shifts, it has to be processed as a table for it to reliably calculate these kpis, I think.